db/snapshot/backup_task.cc

```diff
@@ -0,0 +1,123 @@
+#include "sstables/sstables.hh"
+
+extern logging::logger snap_log;
+
+namespace db::snapshot {
+
+backup_task_impl::backup_task_impl(tasks::task_manager::module_ptr module, snapshot_ctl& ctl, shared_ptr<s3::client> client, sstring bucket, sstring ks, sstring snapshot_name) noexcept
+    : tasks::task_manager::task::impl(module, tasks::task_id::create_random_id(), 0, "node", ks, "", "", tasks::task_id::create_null_id())
+    , _snap_ctl(ctl)
+    , _client(std::move(client))
+    , _bucket(std::move(bucket))
+    , _ks(std::move(ks))
+    , _snapshot_name(std::move(snapshot_name))
+{}
+
+std::string backup_task_impl::type() const {
+    return "backup";
+}
+
+tasks::is_internal backup_task_impl::is_internal() const noexcept {
+    return tasks::is_internal::no;
+}
+
+future<> backup_task_impl::run(sstring data_dir) {
+    return async([this, data_dir = std::move(data_dir)] {
+        gate uploads;
+        auto wait = defer([&uploads] { uploads.close().get(); });
+        std::exception_ptr ex;
+        //
+        // The keyspace data directories and their snapshots are arranged as follows:
+        //
+        //  <data dir>
+        //  |- <keyspace name1>
+        //  |  |- <column family name1>
+        //  |     |- snapshots
+        //  |        |- <snapshot name1>
+        //  |          |- <snapshot file1>
+        //  |          |- <snapshot file2>
+        //  |          |- ...
+        //  |        |- <snapshot name2>
+        //  |        |- ...
+        //  |  |- <column family name2>
+        //  |  |- ...
+        //  |- <keyspace name2>
+        //  |- ...
+        //
+        auto ks_dir = fs::path(data_dir) / _ks;
+        auto ks_dir_lister = directory_lister(ks_dir, lister::dir_entry_types::of<directory_entry_type::directory>());
+        auto close_ks_dir_lister = deferred_close(ks_dir_lister);
+
+        snap_log.trace("Scan keyspace dir {}", ks_dir.native());
+        while (auto table_ent = ks_dir_lister.get().get()) {
+            auto snapshot_dir = ks_dir / table_ent->name / sstables::snapshots_dir / _snapshot_name;
+            auto has_snapshot = file_exists(snapshot_dir.native()).get();
+
+            snap_log.trace("Check table snapshot dir {} (exists={})", snapshot_dir.native(), has_snapshot);
+            if (!has_snapshot) {
+                continue;
+            }
+
+            auto cf_name_and_uuid = replica::parse_table_directory_name(table_ent->name);
+            auto snapshot_dir_lister = directory_lister(snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+            auto close_snapshot_dir_lister = deferred_close(snapshot_dir_lister);
+
+            while (auto component_ent = snapshot_dir_lister.get().get()) {
+                auto gh = uploads.hold();
+                auto component_name = snapshot_dir / component_ent->name;
+                auto destination = fmt::format("/{}/{}/{}/{}", _bucket, cf_name_and_uuid.first, _snapshot_name, component_ent->name);
```

Comment:
Doesn't it make sense to invert the structure and put the snapshot first?

If this is compatible with an existing backup structure, we can consider it later.

The layout should be documented.

Please see https://stackoverflow.com/questions/52443839/s3-what-exactly-is-a-prefix-and-what-ratelimits-apply. I think there isn't a problem here - those are metadata limits and we'll be data bound. For random small reads, we may want a prefix-per-sstable, but it's not needed for backup.

Reply:
We may end up having the explicit `prefix` parameter, so the layout will be up to the caller to define

Reply:
> Doesn't it make sense to invert the structure and put the snapshot first?
> 
> If this is compatible with an existing backup structure, we can consider it later.
> 
> The layout should be documented.
> 
> Please see https://stackoverflow.com/questions/52443839/s3-what-exactly-is-a-prefix-and-what-ratelimits-apply. I think there isn't a problem here - those are metadata limits and we'll be data bound. For random small reads, we may want a prefix-per-sstable, but it's not needed for backup.

Horror stories :)
https://stackoverflow.com/questions/58433594/aws-s3-slowdown-please-reduce-your-request-rate/59479230
Bottom line, randomize everything including your buckets